src/main/scala/is/hail/annotations/Region.scala

```diff
@@ -7,292 +7,186 @@ import com.esotericsoftware.kryo.io.{Input, Output}
     storeByte(b,
       (loadByte(b) | (1 << (bitOff & 7))).toByte)
   }
 
-  def clearBit(byteOff: Long, bitOff: Long) {
+  final def clearBit(byteOff: Long, bitOff: Long) {
     val b = byteOff + (bitOff >> 3)
     storeByte(b,
       (loadByte(b) & ~(1 << (bitOff & 7))).toByte)
   }
 
-  def storeBit(byteOff: Long, bitOff: Long, b: Boolean) {
+  final def storeBit(byteOff: Long, bitOff: Long, b: Boolean) {
     if (b)
       setBit(byteOff, bitOff)
     else
       clearBit(byteOff, bitOff)
   }
 
-  def appendInt(i: Int): Long = {
-    val off = allocate(4, 4)
-    storeInt(off, i)
-    off
-  }
-
-  def appendBoolean(b: Boolean): Long =
-    appendByte(b.toByte)
-
-  def appendLong(l: Long): Long = {
-    val off = allocate(8, 8)
-    storeLong(off, l)
-    off
-  }
-
-  def appendFloat(f: Float): Long = {
-    val off = allocate(4, 4)
-    storeFloat(off, f)
-    off
-  }
-
-  def appendDouble(d: Double): Long = {
-    val off = allocate(8, 8)
-    storeDouble(off, d)
-    off
-  }
-
-  def appendByte(b: Byte): Long = {
-    val off = allocate(1)
-    storeByte(off, b)
-    off
-  }
-
-  def appendBytes(bytes: Array[Byte]): Long = {
-    val off = allocate(bytes.length)
-    storeBytes(off, bytes)
-    off
-  }
-
-  def appendBytes(bytes: Array[Byte], bytesOff: Long, n: Int): Long = {
-    assert(bytesOff + n <= bytes.length)
-    val off = allocate(n)
-    storeBytes(off, bytes, bytesOff, n)
-    off
+  final def appendBinary(v: Array[Byte]): Long = {
+    val len: Int = v.length
+    val grain = if (TBinary.contentAlignment < 4) 4 else TBinary.contentAlignment
+    val addr = allocate(grain, 4+len) + (grain-4)
+    storeInt(addr, len)
+    storeBytes(addr+4, v)
+    addr
   }
-
-  def appendBinary(bytes: Array[Byte]): Long = {
-    align(TBinary.contentAlignment)
-    val startOff = appendInt(bytes.length)
-    appendBytes(bytes)
-    startOff
-  }
-
-  def appendBinarySlice(
+  
+  final def appendBinarySlice(
     fromRegion: Region,
     fromOff: Long,
     start: Int,
-    n: Int
+    len: Int
   ): Long = {
-    assert(fromOff + start + n <= fromRegion.size)
-    assert(n >= 0)
-    align(TBinary.contentAlignment)
-    val off = allocate(4, n + 4)
-    storeInt(off, n)
-    copyFrom(fromRegion, TBinary.bytesOffset(fromOff) + start, off + 4, n)
-    off
+    assert(len >= 0)
+    val grain = if (TBinary.contentAlignment < 4) 4 else TBinary.contentAlignment
+    val addr = allocate(grain, 4+len) + (grain-4)
+    storeInt(addr, len)
+    copyFrom(fromRegion, TBinary.bytesOffset(fromOff) + start, addr+4, len)
+    addr
+  }
+
+  // Use of appendXXX methods is deprecated now that Region uses absolute
+  // addresses and non-contiguous memory allocation.  You can't assume any
+  // relationships between the addresses returned by appendXXX methods -
+  // and to make it even more confusing, there may be long sequences of 
+  // ascending addresses (within a buffer) followed by an arbitrary jump
+  // to an address in a different buffer.
```

Comment:
Let's delete them and choose a semantically clear name for the ones that need to stay around (appendArrayInt, appendString/Binary, etc.)  Maybe allocArrayInt, etc.?

Reply:
My intent is to make the minimal and least-disruptive changes to get to having off-heap
C++ Regions.  The implication is that methods which are not used anywhere can be removed,
but methods which *are* used elsewhere keep the same names.

If we want to change names to make the name more accurately represent the meaning,
then IMO that should be done in a later PR which is pure name-changes and no actual
changes in code behavior.

[And that later PR could also address the no-longer-accurate use of "offset" for what
used to be offset-withing-contiguous-memory-of-Region, but is now "absolute address"].